src/core/eme/session.ts

```diff
@@ -314,174 +304,156 @@ function createSessionAndKeyRequest(
-    sessionType,
-    initDataType,
-    initData,
-    errorStream
-  )
-    .catch((error) => {
-      if (error.code !== ErrorCodes.KEY_GENERATE_REQUEST_ERROR) {
-        throw error;
+function generateRequestOnSession(
+  session: MediaKeySession|IMediaKeySession,
+  encryptedEvent : MediaEncryptedEvent
+): Observable<ISessionEvent> {
+    return Observable.defer(() => {
+
+      if (encryptedEvent.initData == null) {
+        const error = new Error("no init data found on media encrypted event.");
+        throw new EncryptedMediaError("INVALID_ENCRYPTED_EVENT", error, true);
       }
 
-      // TODO In that case, the first in $loadedSessions could be this session,
-      // is this wanted?
-      const firstLoadedSession = $loadedSessions.getFirst();
-      if (!firstLoadedSession) {
-        throw error;
-      }
+      const initData = new Uint8Array(encryptedEvent.initData);
+      const initDataType = encryptedEvent.initDataType;
 
-      log.warn("eme: could not create a new session, " +
-               "retry after closing a currently loaded session", error);
+      const generateKeyRequest$ =
+        generateKeyRequest(session, initData, initDataType, "temporary");
 
-      return $loadedSessions.deleteAndClose(firstLoadedSession)
-        .mergeMap(() =>
-          createSessionAndKeyRequest(
-            mediaKeys, keySystem, sessionType, initDataType,
-            initData, errorStream
-          )
-        );
+        return generateKeyRequest$;
     });
 }
 
-/**
- * Create persistent MediaKeySession and load data from a sessionId.
- * @param {MediaKeys} mediaKeys
- * @param {Object} keySystem
- * @param {string} storedSessionId
- * @param {string} initDataType
- * @param {UInt8Array} initData
- * @param {Subject} errorStream
- * @returns {Observable}
- */
-function createPersistentSessionAndLoad(
-  mediaKeys: IMockMediaKeys|MediaKeys,
-  keySystem: IKeySystemOption,
+export function createOrReuseSessionWithRetry(
+  encryptedEvent: MediaEncryptedEvent,
+  mediaKeysInfos: IMediaKeysInfos
+) {
+  return createOrReuseSession(
+    encryptedEvent,
+    mediaKeysInfos
+  ).catch((error) => {
+    if (error.code !== ErrorCodes.KEY_GENERATE_REQUEST_ERROR) {
+      throw error;
+    }
+    const firstLoadedSession = $loadedSessions.getFirst();
+    if (!firstLoadedSession) {
+      throw error;
+    }
+
+    log.warn("eme: could not create a new session, " +
+      "retry after closing a currently loaded session", error);
+
+    return $loadedSessions.deleteAndClose(firstLoadedSession)
+      .mergeMap(() => {
+        return createOrReuseSession(
+          encryptedEvent,
+          mediaKeysInfos
+        );
+      }
+      );
+  });
+}
+
+function createOrReuseSession(
+  encryptedEvent: MediaEncryptedEvent,
+  mediaKeysInfos: IMediaKeysInfos
+): Observable<ISessionEvent> {
+  // reuse currently loaded sessions without making a new key request
+  const initData = new Uint8Array(encryptedEvent.initData || new ArrayBuffer(0));
```

Comment:
Huh? I think I get why the OR (because you already checked for it), but if that's it we shouldn't lie about it by doing weird things like this.

Maybe we could provide directly the initData and initDataType of the event instead of the event, maybe we should check for its presence here instead of there.

From what I can see now though initData is already always checked in ``eme/index.ts``